File: root_orchestrator/cloud-scheduler/manager_requests.py
import os

import requests
from bson.json_util import dumps

SYSTEM_MANAGER_ADDR = (
    "http://"
    + os.environ.get("SYSTEM_MANAGER_URL")
    + ":"
    + str(os.environ.get("SYSTEM_MANAGER_PORT"))
)


def manager_request(cluster, job_id, job, replicas):
    print("sending scheduling result to system-manager...")
    request_address = SYSTEM_MANAGER_ADDR + "/api/result/deploy"
    print(request_address)
    try:
        requests.post(
            request_address,
            json={"cluster_id": str(cluster.get("_id")), "job_id": job_id},
        )
    except requests.exceptions.RequestException:
        print("Calling System Manager /api/result/deploy not successful.")


def manager_request_replicate(cluster, job_id, job, replicas):
    request_address = SYSTEM_MANAGER_ADDR + "/api/result/replicate"
    try:
        requests.post(
            request_address,
            json=dumps({"cluster": cluster, "job": job, "job_id": job_id, "replicas": replicas}),
        )
    except requests.exceptions.RequestException:
        print("Callig System Manager /api/result/replicate not successful.")

File: root_orchestrator/scheduler/requests/manager/managerRequests.go
// Package manager interfaces with the system or cluster manager
package manager

import (
	"bytes"
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"scheduler/logger"
)

var (
	SYSTEM_MANAGER_URL  = os.Getenv("SYSTEM_MANAGER_URL")
	SYSTEM_MANAGER_PORT = os.Getenv("SYSTEM_MANAGER_PORT")
)

const (
	PROTOCOL = "http"
	DEPLOY   = "/api/result/deploy"
)

type deploymentRequest struct {
	JobId     string `json:"job_id"`
	ClusterId string `json:"cluster_id"`
}

// Deploy sends the scheduling result to the system or cluster manager
func Deploy(jobID string, clusterID string) error {
	url := fmt.Sprintf("%s://%s:%s%s", PROTOCOL, SYSTEM_MANAGER_URL, SYSTEM_MANAGER_PORT, DEPLOY)

	req := deploymentRequest{jobID, clusterID}
	payload, err := json.Marshal(req)
	if err != nil {
		logger.ErrorLogger().Println("Could not marshal deployment request")
		return err
	}

	resp, err := http.Post(url, "application/json", bytes.NewBuffer(payload))
	if err != nil {
		logger.ErrorLogger().Println("Could not send deployment request")
		return err
	}

	err = resp.Body.Close()
	if err != nil {
		return err
	}
	return nil
}
